### PR TITLE
Add resource to allow Opensearch audit config to be updated

### LINKED ISSUE #273: Audit log resource
Hi @phillbaker, am i right in thinking there is no terraform resource that can handle `GET/PUT _plugins/_security/api/audit/config` on as per [AWS docs on audit log API](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/audit-logs.html#audit-log-rest-api) ?

Would you be open to accepting a PR to add this capability?

### PR DESCRIPTION
**Pull request by rm-hull**
Fixes #273 

Here's a passing build: https://github.com/rm-hull/terraform-provider-elasticsearch/actions/runs/2339641867.
The test is only enabled for opensearch 1.x.y.

I've left this in a draft state for now, but a few questions and some steering would be good @phillbaker: 
- ~I haven't put a version check in the actual resource .. do you think that is necessary?~ FIXED
- ~I've modelled the `compliance.read_watched_fields` as `map[string][]string` (i.e. as per what comes back from the API), but unsure how represent this in terraform, and have consequently left the population of this blank - see https://github.com/phillbaker/terraform-provider-elasticsearch/pull/276/files#diff-d5f60cf3d507721c2ffc3fb6d62661dd58c1da2e476da6eb012d17ec8460c013R342~ - refactored this so that there are zero or more `read_watched_field` blocks .. it turns out that you cannot model a `map[string][]string` in Terraform SDK: it only supports primitive types (see https://github.com/hashicorp/terraform-plugin-sdk/blob/19b8cceba7264b9551c2c2a3acf633685e4eec8d/helper/schema/schema.go#L2074-L2114)
- I have followed what has gone before with the flatten/expand structures, but have left them in the specific resource file, rather than putting in utils.

### CODE REVIEW

**Comment by phillbaker** [comment]
Thanks for this @rm-hull! It looks like you ended up pretty close to where my comments would have suggested. 

> I have followed what has gone before with the flatten/expand structures, but have left them in the specific resource file, rather than putting in utils.

Yes, this is what I've been transitioning to as well.

**Comment by ZidCode** [comment]
Hi @phillbaker, 
is there anything preventing the PR from being merged ? 
Best regards,
Joanna

File: es/resource_elasticsearch_audit_config.go [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+	return response, nil
+}
+
+// Response used by the security plugin API (GET method)
+type getResponse struct {
+	Config config `json:"config"`
+}
+
+// Response sent by the security plugin API (PUT method)
+type putResponse struct {
```
Comment (phillbaker) [inline es/resource_elasticsearch_audit_config.go:485]: can we make these names more descriptive so it doesn't conflict with other responses, e.g. getAuditResponse, putAuditResponse, auditConfig, auditCompliance?

**Review by phillbaker — COMMENTED** [review]
Thanks for this, it's looking good. Left a few comments after we resolve those, should be good to merge

File: main.go [hunk 2]
```diff
@@ -20,7 +20,7 @@ func main() {
 	flag.Parse()
 
 	if debugMode {
-		err := plugin.Debug(context.Background(), "registry.terraform.io/providers/phillbaker/elasticsearch",
+		err := plugin.Debug(context.Background(), "registry.terraform.io/phillbaker/elasticsearch",
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply main.go:23]: So this is the root cause of why i couldn't debug the provider with the `TF_REATTACH_PROVIDERS` as the documentation added in the README. Without this change, debugging crashes with the following error:

```console
$ export TF_REATTACH_PROVIDERS='{"registry.terraform.io/providers/phillbaker/elasticsearch":{"Protocol":"grpc","ProtocolVersion":5,"Pid":90013,"Test":true,"Addr":{"Network":"unix","String":"/var/folders/32/3mbbgs9x0r5bf991ltrl3p280000gs/T/plugin2410135162"}}}'

$ terraform apply
Error parsing %!q(PANIC=String method: called String on zero-value addrs.Provider) as a provider address: 
Invalid provider source string: The "source" attribute must be in the format "[hostname/][namespace/]name"
```

I supplied the other PR (#274) which fixes this but looking at the PR description, it's not clear what that attempts to fix (so I will update that). I can take this line out of this PR if you want?

File: es/provider.go [hunk 3]
```diff
@@ -250,6 +250,7 @@ func Provider() *schema.Provider {
 			"elasticsearch_xpack_user":                      resourceElasticsearchXpackUser(),
 			"elasticsearch_xpack_watch":                     resourceElasticsearchXpackWatch(),
 			"elasticsearch_script":                          resourceElasticsearchScript(),
+			"elasticsearch_opensearch_audit_config":         resourceOpenSearchAuditConfig(),
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply es/provider.go]: The resources weren't in sorted order .. well, fixed now in 0c52837

File: examples/resources/elasticsearch_audit_settings/resource.tf [hunk 4]
```diff
@@ -0,0 +1,47 @@
+resource "elasticsearch_opensearch_audit_config" "test" {
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply examples/resources/elasticsearch_audit_settings/resource.tf:1]: Added

File: README.md [hunk 5]
```diff
@@ -286,6 +286,51 @@ Please see [the documentation](./docs/index.md#AWS-authentication) for details.
...
+docker-compose up -d
+docker-compose ps -a
+```
+
+When running tests, ensure that your test/debug profile has environmental variables as below:
+
+- `ELASTICSEARCH_URL=http://localhost:9200_`
+- `TF_ACC=1`
+
+### Running terraform with a local provider
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply README.md:311]: Ah, I guess this should be how to run the provider in a debugger. I'll update the title and add another section about setting the dev overrides

File: es/resource_elasticsearch_audit_config.go [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+	if err := resourceElasticsearchAuditConfigCheckVersion(m); err != nil {
+		return err
+	}
+
+	if _, err := resourceElasticsearchPutAuditConfig(d, m); err != nil {
+		return err
+	}
+
+	// There is no identifier associated with audit config, so just use a random one
+	id, err := uuid.GenerateUUID()
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply es/resource_elasticsearch_audit_config.go]: Good point, updated 👍

**Review by phillbaker — APPROVED** [review]
thanks!

**Pull request merged**